### ISSUE TITLE
【模型微调】LLM微调

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#45]
# 什么是微调

微调的对象是模型的参数，常见的方法有：

1. full-parameters fine-tune，全参数微调，会调整模型的所有参数，相对来说更加耗费资源。
2. paramater-efficient fine-tune（PEFT），参数高效微调，只调整模型的一部分参数或添加少量的新参数。比较常见的是：
    1. LoRA（Low-Rank Adaption）：将模型权重矩阵分解为低秩矩阵，仅训练降维后的参数，从而减少需要调整的参数数量。
    2. Prompt Tuning：通过在输入中添加可训练的提示向量，引导模型生成所需输出，而不改变模型本身的参数。
3. 指令调优（Instruction Tuning）：通过使用针对特定任务设计的指令进行微调，使模型更好地理解和执行用户指令。
4. 强化学习微调（Reinfocement Learning Fine-Tuning），分为：    
    1. RLHF（Reinforcement Learning from Human Feedback）：结合人类反馈，通过强化学习的方法优化模型的输出，使其更符合人类期望。
    2. 其他强化学习方法（如策略梯度、Q-learning等）也可用于微调LLM
5. 领域自适应微调（Domain Adaptive Fine-Tuning）
6. 低资源微调技术：针对低资源环境（如计算能力有限或数据稀缺的情况），开发了多种技术以实现高效微调。包括：
    1. 知识蒸馏（Knowledge Distillation）：通过将大型LLM的知识转移到较小的学生模型，使得学生模型在资源受限的环境中也能表现良好
    2. 量化（Quantization）：通过将模型权重从高精度（如32位浮点）转换为低精度表示（如8位整数），减少模型的存储和计算需求。

huggingface 有 PEFT 的开源实现框架：https://github.com/huggingface/peft，微调方法实现了 LoRA、Prompt Tuning 和 IA3. [^2]

# 有哪些常见的微调方法

根据更新参数的数量，分为两类：

1. 微调时会更新全量参数，称为全量微调（Full Fine-Tuning），在任务中会更新预训练模型的所有参数。这种方法通常效果好，但开销大，对存储和计算资源要求高，同时在多个任务之间切换时需要维护每个任务的完整模型参数。

2. 参数高效的微调方法，不会更新全量参数。常见的有：
    1. 层冻结微调（layer Freezing / Partial Fine-tuning）：
    2. Adaptor微调：
    3. Prompt Tuning / Prefix Tuning：
    4. LoRA（Low-Rank Adaption）：
    5. BitFit（仅微调偏置参数）：xxx

# 常见的微调工具

1. Huggingface Transformers：目前最流行的预训练模型库之一，支持多种架构（如BERT、GPT、T5等），并提供简单易用的接口。内置了 Trainer API，方便快速微调，同时也支持分布式训练和多卡加速。此外，其生态系统中还衍生出了许多扩展，比如 Adaptor Transformers。https://github.com/huggingface/transformers
2. Huggingface PEFT（Parameter-Efficient Fine-Tuning）：针对参数高效微调设计的专门工具库，整合了如 LoRA、Prefix Tuning、Adaptor 等多种微调方法。使用 PEFT 可以在资源受限的情况下对超大模型快速进行微调。https://github.com/huggingface/peft
3. AdaptorHub：AdaptorHub 是构建于Transformers 库基础上的一个扩展，专注于adaptor 模块的训练和管理。用户可以方便地在预训练模型中插入、训练和复用adaptor，使得多任务、多领域迁移变得简单高效。仓库：https://github.com/adapter-hub/adapters
4. DeepSpeed：微软开发的 DeepSpeed 不仅针对分布式训练和大规模模型训练进行了优化，同时也提供了如 ZeRO 优化器等内存优化方法，帮助在有限的资源下对大模型进行微调。DeepSpeed 还支持一些参数高效微调方案，适用于超大规模模型。https://github.com/deepspeedai/DeepSpeed
5. PyTorch Lightning：作为PyTorch 的高级封装框架，Lightning简化了训练循环、日志记录、分布式训练等众多繁琐细节。对于需要自定义复杂微调逻辑的场景，Lightning能帮助开发者更快地搭建训练框架，并且易于扩展和调试。https://github.com/Lightning-AI/pytorch-lightning
6. TensorFlow/Keras 生态下的工具：对于TensorFlow 用户，Keras提供了简单的API用于构建、微调模型。另外，TensorFlow Hub上也有许多预训练模型，用户可以直接加载后再进行下游任务微调。https://github.com/keras-team/keras
7. Fairseq：这是Facebook AI Research 推出的一个用于序列到序列任务的工具包，主要应用于翻译、文本生成等任务。虽然Fairseq较为侧重于研究领域和前沿模型，但其中也包含了不少用于微调的工具和示例代码。https://github.com/facebookresearch/fairseq

小结论：如果资源受限，用Huggingface PEFT、DeepSpeed。

# 如何准备数据

## 数据准备步骤

1. 数据清洗和预处理：
    - 确保数据质量：去除特殊符号、无效字符或格式问题等噪音信息
    - 格式统一：整理成统一的结构（如CSV、JSON或文本文件），便于后续处理和加载

2. Tokenization（分词）
    - 使用与预训练模型一致的分词器：大多数工具和库（如Huggingface Transformers）都提供了预置的分词器，使用同一个分词器（如BERT的WordPiece、GPT的Byte-Pair Encoding）可以确保输入的token id与模型预训练时的词表保持一致
    - 注意截断和填充：为保证输入在同一长度，通常需要设置最大序列长度，并对过长文本进行截断，对不足长度的文本进行填充（padding）。

3. 使用数据加载工具和流水线
    - Huggingface Datasets：这个库可以快速加载、处理和转换数据集，支持分布式加载。可以利用 map 函数批量执行分词、清洗以及格式转换。https://github.com/huggingface/datasets

4. 数据增强和采样策略（可选）
    - 数据增强：在数据量较少的场景下，可以尝试使用数据增强方法（如同义词替换、随机删除等）以增加数据多样性
    - 采样平衡：对于类别不平衡的任务，可能需要设计采样策略来避免模型训练时偏向于高频类别

## 数据格式

不同的微调任务对于输入和输出的数据格式都有一些差异，常见的数据格式包括：

1. 指令-回复：每个样本包含 "instruction" 和 "response" 字段

```
{
  "instruction": "请介绍一下太阳系行星。",
  "response": "太阳系有八大行星..."
}
```

2. 问答格式（Question-Answer）：每个样本都包含一个问题字段和一个答案字段，适用于事实性问答任务

```
{
"question": "中国的首都是哪里？",
"answer": "北京"
}
```

3. 多轮对话（Multi-turn Conversation）格式: 这种格式记录了对话历史，有利于训练能理解上下文的对话系统。

```
{
"conversation": [
{"speaker": "用户", "message": "你好！"},
{"speaker": "助手", "message": "您好，我是您的智能助手。"},
{"speaker": "用户", "message": "今天天气怎么样？"},
{"speaker": "助手", "message": "据最新信息，今天晴朗，温度适宜。"}
]
}
```

4. 指令、示例与回复（Instruction with Demonstrations）：在指令之外，额外提供了一到多个示例，帮助模型理解任务，然后对新的输入给出适当回复

```
{
"instruction": "将以下英文句子翻译成中文。",
"demonstrations": [
{"input": "Good morning.", "output": "早上好。"},
{"input": "How are you?", "output": "你好吗？"}
],
"input": "I love programming.",
"response": "我热爱编程。"
}
```

5. 文本续写（Text Continuation / Completion）:此格式为生成任务设计，模型基于给定的上下文生成后续内容或续写文本

```
{
"context": "在一个遥远的星球上，有一座神秘的城市，",
"continuation": "其中居住着与世隔绝的智慧生物，他们掌握着古老的秘密……"
}
```

6. 分类和标注格式: 针对分类任务，样本除了文本外还标注了该文本对应的类别；对序列标注任务，则提供文本中的各个实体及其类别信息。

```
示例（情感分类）：
{
"text": "这部电影非常精彩！",
"label": "正面"
}
示例（实体识别）：
{
"text": "马云创办了阿里巴巴。",
"entities": [
{"entity": "马云", "type": "人名"},
{"entity": "阿里巴巴", "type": "组织"}
]
}
```

7. 对比学习格式: 该格式适合对比学习任务，通过标注两个文本是否具有相似的语义或观点，来帮助模型区分细微差别.

```
{
"text1": "这款手机的电池续航时间非常长。",
"text2": "这部手机充电速度很快。",
"label": "不同"
}
```

8. 多任务混合格式: 通过在样本中明确标记任务类型（例如“翻译”、“摘要”、“问答”等），使得一个模型能够根据任务标识调整生成策略，同时也可以在同一批数据中混合不同任务进行训练

```
{
"task": "翻译",
"input": "Good night.",
"output": "晚安。"
}
```

# 编写程序

场景：使用 Huggingface PEFT和Datasets库微调 Qwen 2.5 模型

todo

example: train [^1]

# 测评结果

首先准备评测数据，结构和训练数据类似：

```
[
  {
    "instruction": "xxxx",
    "output": "xxx"
  },
  {
    "instruction": "xxxx",
    "output": "xxx"
  }
]
```

编写python程序来测评，主要的思路是比对模型的输出和评测数据中的 output的对比，不同的场景，output对比的方法也不一样：对于RAG等系统，不能用==这种比较，得是判断语义相似性。

[^1]: https://hf-mirror.com/docs/transformers/training#train-in-native-pytorch
[^2]: https://huggingface.co/docs/peft/task_guides/prompt_based_methods
